src/object/clone.ts

```diff
@@ -26,48 +26,103 @@
   }
 
   if (obj instanceof RegExp) {
-    return new RegExp(obj.source, obj.flags) as T;
+    return new RegExp(obj.source, obj.flags) as Shallowed<T>;
   }
 
   if (obj instanceof Map) {
     const result = new Map();
     for (const [key, value] of obj) {
       result.set(key, value);
     }
-    return result as T;
+    return result as Shallowed<T>;
   }
 
   if (obj instanceof Set) {
     const result = new Set();
     for (const value of obj) {
       result.add(value);
     }
-    return result as T;
+    return result as Shallowed<T>;
   }
 
-  if (typeof obj === "object") {
-    return Object.assign({}, obj) as T;
+  // BINARY DATA
+  if (obj instanceof Uint8Array) return new Uint8Array(obj) as Shallowed<T>;
+  if (obj instanceof Uint8ClampedArray) return new Uint8ClampedArray(obj) as Shallowed<T>;
+  if (obj instanceof Uint16Array) return new Uint16Array(obj) as Shallowed<T>;
+  if (obj instanceof Uint32Array) return new Uint32Array(obj) as Shallowed<T>;
+  if (obj instanceof BigUint64Array) return new BigUint64Array(obj) as Shallowed<T>;
+  if (obj instanceof Int8Array) return new Int8Array(obj) as Shallowed<T>;
+  if (obj instanceof Int16Array) return new Int16Array(obj) as Shallowed<T>;
+  if (obj instanceof Int32Array) return new Int32Array(obj) as Shallowed<T>;
+  if (obj instanceof BigInt64Array) return new BigInt64Array(obj) as Shallowed<T>;
+  if (obj instanceof Float32Array) return new Float32Array(obj) as Shallowed<T>;
+  if (obj instanceof Float64Array) return new Float64Array(obj) as Shallowed<T>;
+  if (obj instanceof ArrayBuffer) return obj.slice(0) as Shallowed<T>;
+  if (obj instanceof SharedArrayBuffer) return obj.slice(0) as Shallowed<T>;
+  if (obj instanceof DataView) return new DataView(obj.buffer.slice(0)) as Shallowed<T>;
+  if (obj instanceof File) return new File([obj], obj.name, { type: obj.type }) as Shallowed<T>;
+  if (obj instanceof Blob) return new Blob([obj], { type: obj.type }) as Shallowed<T>;
+
+  if (typeof obj === 'object') {
+    return Object.assign({}, obj) as Shallowed<T>;
   }
-  return obj;
+  return obj as Shallowed<T>;
 }
 
+export type Shallowed<T> = Equal<T, ShallowMain<T>> extends true ? T : ShallowMain<T>;
+type ShallowMain<T> = T extends [never]
+  ? never
+  : T extends object
+    ? T extends
+        | Array<any>
+        | Set<any>
+        | Map<any, any>
+        | Date
+        | RegExp
+        | Date
+        | Uint8Array
+        | Uint8ClampedArray
+        | Uint16Array
+        | Uint32Array
+        | BigUint64Array
+        | Int8Array
+        | Int16Array
+        | Int32Array
+        | BigInt64Array
+        | Float32Array
+        | Float64Array
+        | ArrayBuffer
+        | SharedArrayBuffer
+        | DataView
+        | Blob
+        | File
+      ? T
+      : OmitNever<{
+          [P in keyof T]: T[P] extends Function ? never : T[P];
+        }>
+    : T;
+
+type Equal<X, Y> = X extends Y ? (Y extends X ? true : false) : false;
```

Comment:
@samchon 

```ts
type Equal<X, Y> = X extends Y ? (Y extends X ? true : false) : false;
type Answer = Equal<1 | 2, 1 | 2>; // It infer boolean. I think it is weird.
```

If it is not wrapped in generic again, the union type branches into each element type to ride the conditional type, so if the two union types are the same, it is inferred to be boolean. For complete reasoning about the union type, the Equal type must be modified.

```ts
export type Equal<X, Y> =
    (<T>() => T extends X ? 1 : 2) extends
    (<T>() => T extends Y ? 1 : 2) ? true : false;

type Answer = Equal<1 | 2, 1 | 2>; // true
```

Reply:
Good